Fix mbedTLS missing TLS v1.3 support with OpenSSL

ynezz opened the pull request:
Current default official images on all devices are using mbedTLS 2.y which
doesn't support TLS v1.3 (#12874), so lets fix it by using OpenSSL on devices
which should have enough storage space to accomodate additional ~1.5MB
(rough estimate didn't measured it).

This is just my first quick attempt to implement one of the possible
solutions, since I think, that using mbedTLS 3.y is not an option for backport
into 23.05.

While at it, make the dash/underscore naming pattern in `FEATURES` consistent
and use dash, cleanup apparently unused features as well.

ynezz committed:
build: add GRAND_FLASH feature

So we can mark targets which have enough flash space to accomodate
certain features.

Signed-off-by: Petr Štetiar <ynezz@true.cz>

ynezz committed:
build: add WIRELESS_SUPPORT feature

So we can clearly mark targets having wireless support.

Signed-off-by: Petr Štetiar <ynezz@true.cz>

ynezz committed:
treewide: unify wireless feature handling

Currently its a quite common pattern to include `wpad-basic-mbedtls`
package when the device/target should've wireless support, hardcoding
TLS provider as well, making any changes quite cumbersome.

Lets make the whole wireless/TLS situation more flexible and start using
new wireless target specific feature and explicitly disable missing
wireless support with new opaque $(PACKAGE_NO_WIRELESS) variable which
is going to remove the wireless packages from devices which lack
wireless support, so currently PACKAGE_NO_WIRELESS is essentially
`-wpad-basic-mbedtls`.

This way we should've complete control over wireless/TLS features in a
single target.mk place, thus making future adjustements more flexible.

This change shouldn't have any side effects, its just reworking current
device package sets with the new wireless/TLS feature handling.

Signed-off-by: Petr Štetiar <ynezz@true.cz>

ynezz committed:
target: grand flash devices should use OpenSSL

During the years we've learned it hard way, that we needed to make a lot
of compromises while using flash space friendly micro TLS libraries like
mbedTLS/wolfSSL in order to provide more or less up to date security
features on most supported devices.

Most of the recent and decent devices have plenty of storage space, so
there is no need to make such compromises anymore and we could simply
use battle tested OpenSSL on such targets by default as storage space
increase is around 1.5 MiB, which is no brainer.

So lets make it possible to use OpenSSL on grand flash devices and
switch to libustream-openssl and wpad-basic-openssl by default there.

This should have no functional change, the target needs to actually
explicitly define `FEATURES := grand-flash` in order to have OpenSSL by
default.

References: #12874
Signed-off-by: Petr Štetiar <ynezz@true.cz>

ynezz committed:
treewide: mark grand-flash targets

Select targets which should've plenty of storage to enable additional
features like OpenSSL by default for example.

References: #12874
Signed-off-by: Petr Štetiar <ynezz@true.cz>

ynezz committed:
scan.mk: do not silence output of dump phase

Make it easier to spot currently hidden issues:

 $ make defconfig V=sc
 ...
 Collecting target info: target/linux/airohabash: -c: line 1: syntax error near unexpected token `|'
 bash: -c: line 1: `echo 1686815253<LINUX_VERMAGIC> | staging_dir/host/bin/mkhash md5 | cut -b1-8'
 bash: -c: line 1: syntax error near unexpected token `|'
 bash: -c: line 1: `echo 1686815253<LINUX_VERMAGIC> | staging_dir/host/bin/mkhash md5 | sed -E 's/(.{8})(.{4})(.{4})(.{4})(.{10})../\1-\2-\3-\4-\500/''
 ...

Signed-off-by: Petr Štetiar <ynezz@true.cz>

ynezz committed:
treewide: fix shell errors during dump stage

Fixes following issues:

 bash: -c: line 1: `echo 1686820180<LINUX_VERMAGIC> | /staging_dir/host/bin/mkhash md5 | cut -b1-8'
 bash: -c: line 1: `echo 1686820180<LINUX_VERMAGIC> | /staging_dir/host/bin/mkhash md5 | sed -E 's/(.{8})(.{4})(.{4})(.{4})(.{10})../\1-\2-\3-\4-\500/''
 bash: -c: line 1: syntax error near unexpected token `|'
 bash: line 1: *1024*1024: syntax error: operand expected (error token is "*1024*1024")
 bash: line 1: (64 + ): syntax error: operand expected (error token is ")")
 expr: syntax error: missing argument after '+'

Signed-off-by: Petr Štetiar <ynezz@true.cz>

Ansuel commented:
@ynezz it was pointed out on IRC that maybe BIG might be a better naming and more descriptive?

ynezz commented:
> @ynezz it was pointed out on IRC that maybe BIG might be a better naming and more descriptive? 

Well, I know only [Grandmaster Flash](https://en.m.wikipedia.org/wiki/Grandmaster_Flash) :)

ynezz committed:
target: sunset jffs2_nand feature

Seems to be unused, the only user was removed in commit 44447f688209
("kirkwood: clean up FEATURES") with "Legacy build code has been
removed, NAND devices are only supported with UBI now" reasoning.

Signed-off-by: Petr Štetiar <ynezz@true.cz>

ynezz committed:
target: transition feature low_mem to low-mem

To make the naming scheme consistent.

Signed-off-by: Petr Štetiar <ynezz@true.cz>

ynezz committed:
target: transition feature separate_ramdisk to separate-ramdisk

To make the naming scheme consistent.

Signed-off-by: Petr Štetiar <ynezz@true.cz>

ynezz committed:
target: transition feature small_flash to small-flash

To make the naming scheme consistent.

Signed-off-by: Petr Štetiar <ynezz@true.cz>

ynezz committed:
target: sunset spe_fpu feature

Seems to be unused.

Signed-off-by: Petr Štetiar <ynezz@true.cz>

raenye commented on target/linux/apm821xx/image/sata.mk:
```diff
@@ -9,7 +9,7 @@ define Device/wd_mybooklive
   DEVICE_ALT0_MODEL := My Book Live Duo
   DEVICE_PACKAGES := kmod-usb-dwc2 kmod-ata-dwc kmod-usb-ledtrig-usbport \
-	kmod-usb-storage kmod-fs-vfat wpad-basic-mbedtls
+	kmod-usb-storage kmod-fs-vfat
```
Perhaps `$(PACKAGE_NO_WIRELESS)` fits here too? MBL and MBLD are wired devices.

ynezz replied:
Thanks for the review, [should be fixed](https://github.com/openwrt/openwrt/pull/12917/commits/97d2b66ba1de921b9c602b1dcfbd212e68d0a90c).

raenye commented:
Does it make sense to also make `libcurl` use the default TLS implementation?

ynezz committed:
apm821xx: wd_mybooklive: remove wireless support

MBL are wired devices.

References: https://github.com/openwrt/openwrt/pull/12917#discussion_r1231529208
Suggested-by: Rani Hod <rani.hod@gmail.com
Signed-off-by: Petr Štetiar <ynezz@true.cz>

Shine- commented:
I'd call it "LARGE_FLASH", not "GRAND_FLASH".

One more suggestion: In 23.05, the `luci-ssl` package has switched dependencies from WolfSSL to MBEDTLS, but there's no replacement package providing LuCI with a WolfSSL backend. So users with smaller flash currently have no possibility to get back TLS 1.3 support, by e.g. using WolfSSL.

Therefore, it might be beneficial to provide a new package `luci-ssl-wolfssl` similar to `luci-ssl-openssl` for these users.

ynezz commented:
Yes, naming is hard, so I simply got a list of `small` antonyms and found `grand` to match well the intended use cases, so I went with that.

> BIG might be a better naming and more descriptive

That is highly subjective. Is 32MiB really big? I doubt that, yet 32MiB would still be capable to accommodate for OpenSSL and SELinux for example.

> I'd call it "LARGE_FLASH", not "GRAND_FLASH".

Ok, but the storage does not need to be large, it should be just _sufficient_ for planned use cases within that release(s). I've rejected `sufficient_flash` as it's long and error/typo prone, same with `generous_flash`. I kind of liked `premium_flash` but it sounded like paid feature :)

Djfe commented:
> One more suggestion: In 23.05, the `luci-ssl` package has switched dependencies from WolfSSL to MBEDTLS, but there's no replacement package providing LuCI with a WolfSSL backend. So users with smaller flash currently have no possibility to get back TLS 1.3 support, by e.g. using WolfSSL.
> 
> Therefore, it might be beneficial to provide a new package `luci-ssl-wolfssl` similar to `luci-ssl-openssl` for these users.

wolfssl was dropped again for a reason.
https://openwrt.org/releases/23.05/notes-23.05.0-rc1#switch_from_wolfssl_to_mbedtls_as_default
I'm not sure how much effort it is to provide support for LuCI and how many users benefit, if the package isn't updated often because of ABI changes.

Some discussion on the matter:
https://github.com/openwrt/openwrt/pull/11626#issuecomment-1416258187 (this and the following comments)

Shine- commented:
I understand the WolfSSL ABI part (was hit by it more than once myself) and fully agree with dropping WolfSSL as default library. I also fully agree with making EMBEDTLS the default. In fact, I was hoping for that for years already! :-)

BUT by only providing EMBEDTLS (no TLS 1.3 support as of yet) and OpenSSL (too large for small flash devices), you're completely blocking small-flash devices from having state-of-the-art TLS support. Providing WolfSSL as an _option_ (not default!) would fix that. Apart from that, it's a tiny change anyway - copying the `Makefile` to a new package and updating the `LUCI_DEPENDS` line should do it.

Then, everybody who needs it for their small-flash device could get a custom, WolfSSL-equipped image from the Image Builder.

Just my 2ct anyway.

orangepizza commented:
> I'm aware of the WolfSSL ABI matter (was hit by it more than once myself) and fully agree with dropping WolfSSL as default library. I also fully agree with making EMBEDTLS the default. In fact, I was hoping for that for years already! :-)
> 
> BUT by only providing EMBEDTLS (no TLS 1.3 support as of yet) and OpenSSL (too large for small flash devices), you're completely blocking small-flash devices from having state-of-the-art TLS support. Providing WolfSSL as an _option_ (not default!) would fix that. Apart from that, it's a tiny change anyway - copying the `Makefile` to a new package and updating the `LUCI_DEPENDS` line should do it.
> 
> Then, everybody who needs it for their small-flash device (or for any other reason they might have) could get a custom, WolfSSL-equipped image from the Image Builder.
> 
> Just my 2ct anyway.

well, other option is update mbedtls 3.x, which we must do anyways in next year as 2.28 branch is EOL in 2024 dec

Djfe commented:
maybe it's possible to offer mbedtls 3 instead of wolfssl as an optional luci-ssl package?
and then towards the end of the year after some testing in main, we can switch over fully to version 3?

Gingernut1978 commented:
With this commit applied on latest master menuconfig forces you to use wpad-basic-openssl and won't let you switch to any other wirelessapd openssl option.

ynezz commented:
> won't let you switch to any other wirelessapd openssl option

You should change your preferred TLS provider first in menu `Global build settings` -> `TLS provider`

![image](https://github.com/openwrt/openwrt/assets/43848/c678e7fe-eeed-401d-83a9-387107a51c56)

Gingernut1978 commented:
> > won't let you switch to any other wirelessapd openssl option
> 
> @Gingernut1978 You should change your preferred TLS provider first in menu `Global build settings` -> `TLS provider`
> 
> ![image](https://user-images.githubusercontent.com/43848/246600183-c678e7fe-eeed-401d-83a9-387107a51c56.png)

OK, but it still won't let you switch from the wpad-basic.

ynezz commented:
> OK, but it still won't let you switch from the wpad-basic package.

@Gingernut1978 try the latest update, should be fixed by using `imply` weak version of `select` keyword.

raenye commented:
> Does it make sense to also make `libcurl` use the default TLS implementation?

@ynezz (maybe you missed this)

Also, is the wireless-support flag accessible from LuCI? so wireless features are suppressed by default in, e.g., dashboard or statistics.

ynezz commented:
BTW everything should stay working as it is today, just bunch of devices would use fully featured OpenSSL, thats simply all about this PR. Everything else is just some refactoring to make the handling of wireless/TLS features easier.

> Does it make sense to also make `libcurl` use the default TLS implementation?
 
I don't plan to look into that, but I assume, that some new config options might make such handling easier.

> Also, is the wireless-support flag accessible from LuCI? so wireless features are suppressed by default in, e.g., dashboard or statistics.

I assume, that LuCI is using some kind of runtime probing to guess available features, since it works today as well, without any such specific config option.

orangepizza commented:
I'm trying to update to mbedtls3 buy as we stripped versions from building have hard time to differentiate 2 and 3 
I;m using a deprecated const to test if it's 2 or 3, but not sure if its not brittle enought

ynezz commented:
> I'm trying to update to mbedtls3 buy as we stripped versions from building have hard time to differentiate 2 and 3 I;m using a deprecated const to test if it's 2 or 3, but not sure if its not brittle enought

Ok, nice, but would it be please possible to move such unrelated comments into the appropriate PR? In other words, please try to stay on topic, this is about OpenSSL by default, not about curl, LuCI or mbedTLS v3 support.

Gingernut1978 commented:
> > OK, but it still won't let you switch from the wpad-basic package.
> 
> @Gingernut1978 try the latest update, should be fixed by using `imply` weak version of `select` keyword.

Confirmed fixed after latest update. 

Thanks.

Neustradamus commented:
To follow

ynezz commented:
> To follow

@Neustradamus no need to clutter the issue with such posts, check https://docs.github.com/en/account-and-profile/managing-subscriptions-and-notifications-on-github/setting-up-notifications/about-notifications#subscription-options and check notification settings on the right side

![image](https://github.com/openwrt/openwrt/assets/43848/c76202d5-9138-4236-b836-97c850396656)

kwiateusz commented:
I checked the sysupgrade sizes for various combinations and here are the results:

Config for all tests:
```
CONFIG_TARGET_qualcommax=y
CONFIG_TARGET_qualcommax_ipq807x=y
CONFIG_TARGET_qualcommax_ipq807x_DEVICE_xiaomi_ax3600=y
```
Results
```
Type      Size           Additional Flags
mbedTLS	: 11771kB 11.4MB (CONFIG_TLS_PROVIDER_MBEDTLS=y)
mixed	: 12911kB 12.6MB (CONFIG_TLS_PROVIDER_OPENSSL=y)
OpenSSL : 12911kB 12.6MB (CONFIG_TLS_PROVIDER_OPENSSL=y CONFIG_LUA_ECO_OPENSSL=y)
```

```
+curl
Type      Size           Additional Flags
mbedTLS	: 12011kB 11.7MB (CONFIG_TLS_PROVIDER_MBEDTLS=y CONFIG_PACKAGE_curl=y)
mixed	: 13351kB 13.0MB (CONFIG_TLS_PROVIDER_OPENSSL=y CONFIG_PACKAGE_curl=y)
OpenSSL : 13171kB 12.8MB (CONFIG_TLS_PROVIDER_OPENSSL=y CONFIG_LUA_ECO_OPENSSL=y CONFIG_PACKAGE_curl=y CONFIG_LIBCURL_OPENSSL=y )
```

From what I can see, it will be necessary to make corrections in the external packages.

By default `CONFIG_LUA_ECO_MBEDTLS=y` flag is added from packages/lang/lua-eco package

And when curl is selected, it adds `CONFIG_LIBCURL_MBEDTLS=y` by default due to the fact that it is the default option of this library

The image itself on any version works fine as long as you can change the library for encryption in the package.

rmandrad commented:
can someone fix the confusing title for this PR?  Openssl supports TLS v1.3 but the issue is with mbedTLS ...probably something like "Fix mbedTLS  missing TLS v1.3 support with OpenSSL" ?

Ansuel commented:
@ynezz to light up this pr can i merge some of the intermediate one and the sunset part?

Headcrabed commented on include/target.mk:
```diff
@@ -38,10 +38,30 @@ DEFAULT_PACKAGES+=procd-ujail
+DEFAULT_PACKAGES+=libustream-openssl
+PACKAGE_NO_WIRELESS:=-wpad-basic-openssl
+  ifneq ($(CONFIG_WIRELESS_SUPPORT),)
+    DEFAULT_PACKAGES+=wpad-basic-openssl
```
I think we could use wpad-openssl rather than wpad-basic-openssl on grand flash targets. This would enable more features such as 802.11v/r.

ynezz replied:
Ok, once this gets merged feel free to propose that change via separate pull request/patch.

hauke commented on include/target.mk:
```diff
@@ -38,10 +38,30 @@ DEFAULT_PACKAGES+=procd-ujail
 # mbedTLS wireless features handling
+ifeq ($(CONFIG_TLS_PROVIDER_MBEDTLS),y)
 DEFAULT_PACKAGES+=libustream-mbedtls
 PACKAGE_NO_WIRELESS:=-wpad-basic-mbedtls
```
The indention looks strange here, I think you should indent everything inside the ifeq block.

ynezz replied:
Its probably highly subjective :), but I've changed that.

ynezz commented:
> You can't comment at this time — your comment is too long (maximum is 65536 characters). 

So https://gist.github.com/ynezz/4372b7900bb426ee23aa5d47c36cf45a contains config diffs

raenye commented:
I'm not sure dropping `libustream-xxx` for wired devices is right. Did you try this on any wired device?

Obviously `wpad-basic-xxx` and `iwinfo` are only needed with wireless, but AFAICT removing `libustream` cuts  `opkg` network access.
(I once had to fix it by `scp`-ing the `libustream` package to `/tmp/` and installing from there.)

ynezz commented:
Thanks for helping reviewing this, really appreciate that!

> I'm not sure dropping `libustream-xxx` for wired devices is right. Did you try this on any wired device?

BTW where do you see it dropped? FYI that diff might be misleading:

<details>
<summary>libustream config options in TLS enabled provider configs (<i>click to expand</i>)</summary>

```
config-tls-apm821xx_nand CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-apm821xx_sata CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-armsr_armv7 CONFIG_PACKAGE_libustream-openssl=y
config-tls-armsr_armv8 CONFIG_PACKAGE_libustream-openssl=y
config-tls-at91_sam9x CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-at91_sama5 CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-at91_sama7 CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-ath79_generic CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-ath79_mikrotik CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-ath79_nand CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-ath79_tiny CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-bcm27xx_bcm2708 CONFIG_PACKAGE_libustream-openssl=y
config-tls-bcm27xx_bcm2709 CONFIG_PACKAGE_libustream-openssl=y
config-tls-bcm27xx_bcm2710 CONFIG_PACKAGE_libustream-openssl=y
config-tls-bcm27xx_bcm2711 CONFIG_PACKAGE_libustream-openssl=y
config-tls-bcm47xx_generic CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-bcm47xx_legacy CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-bcm47xx_mips74k CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-bcm4908_generic CONFIG_PACKAGE_libustream-openssl=y
config-tls-bcm53xx_generic CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-bmips_bcm6318 CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-bmips_bcm63268 CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-bmips_bcm6328 CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-bmips_bcm6358 CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-bmips_bcm6362 CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-bmips_bcm6368 CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-gemini_generic CONFIG_PACKAGE_libustream-openssl=y
config-tls-imx_cortexa7 CONFIG_PACKAGE_libustream-openssl=y
config-tls-imx_cortexa9 CONFIG_PACKAGE_libustream-openssl=y
config-tls-ipq40xx_chromium CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-ipq40xx_generic CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-ipq40xx_mikrotik CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-ipq806x_chromium CONFIG_PACKAGE_libustream-openssl=y
config-tls-ipq806x_generic CONFIG_PACKAGE_libustream-openssl=y
config-tls-kirkwood_generic CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-lantiq_ase CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-lantiq_xway CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-lantiq_xway_legacy CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-layerscape_armv7 CONFIG_PACKAGE_libustream-openssl=y
config-tls-layerscape_armv8_64b CONFIG_PACKAGE_libustream-openssl=y
config-tls-malta_be CONFIG_PACKAGE_libustream-openssl=y
config-tls-mediatek_filogic CONFIG_PACKAGE_libustream-openssl=y
config-tls-mediatek_mt7622 CONFIG_PACKAGE_libustream-openssl=y
config-tls-mediatek_mt7623 CONFIG_PACKAGE_libustream-openssl=y
config-tls-mediatek_mt7629 CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-mpc85xx_p1010 CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-mpc85xx_p1020 CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-mpc85xx_p2020 CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-mvebu_cortexa53 CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-mvebu_cortexa72 CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-mvebu_cortexa9 CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-mxs_generic CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-octeon_generic CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-octeontx_generic CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-oxnas_ox820 CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-pistachio_generic CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-qualcommax_ipq807x CONFIG_PACKAGE_libustream-openssl=y
config-tls-ramips_mt7620 CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-ramips_mt7621 CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-ramips_mt76x8 CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-ramips_rt288x CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-ramips_rt305x CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-ramips_rt3883 CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-realtek_rtl838x CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-realtek_rtl839x CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-realtek_rtl930x CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-realtek_rtl931x CONFIG_PACKAGE_libustream-mbedtls=y
config-tls-rockchip_armv8 CONFIG_PACKAGE_libustream-openssl=y
config-tls-sifiveu_generic CONFIG_PACKAGE_libustream-openssl=y
config-tls-sunxi_cortexa53 CONFIG_PACKAGE_libustream-openssl=y
config-tls-sunxi_cortexa7 CONFIG_PACKAGE_libustream-openssl=y
config-tls-sunxi_cortexa8 CONFIG_PACKAGE_libustream-openssl=y
config-tls-tegra_generic CONFIG_PACKAGE_libustream-openssl=y
config-tls-x86_64 CONFIG_PACKAGE_libustream-openssl=y
config-tls-x86_generic CONFIG_PACKAGE_libustream-openssl=y
config-tls-x86_geode CONFIG_PACKAGE_libustream-openssl=y
config-tls-x86_legacy CONFIG_PACKAGE_libustream-openssl=y
config-tls-zynq_generic CONFIG_PACKAGE_libustream-openssl=y
```
</details>

> also, apm821xx/sata, which is WD MyBookLive if I understand correctly, still shows wireless support in the config diff above.

Indeed, there are still some weird targets like apm821xx, lantiq, kirkwood to name a few.

ynezz commented:
> EDIT: also, `apm821xx/sata`, which is WD MyBookLive if I understand correctly, still shows wireless support in the config diff above.

Should be handled now properly.

Bellow you can find config diffs for all targets and thus see changes introduced by this pull request on commit [b60dc68ae65f](https://github.com/openwrt/openwrt/pull/12917/commits/b60dc68ae65f)

  <details>
    <summary><b>apm821xx/nand</b> config diff (<i>click to expand</i>)</summary>


```shell
diff -u config-master-apm821xx_nand config-tls-apm821xx_nand
```

```diff
--- config-master-apm821xx_nand	2023-07-28 16:52:18.434312203 +0200
+++ config-tls-apm821xx_nand	2023-07-28 16:52:24.614373390 +0200
@@ -100,8 +100,6 @@
 CONFIG_MODULE_DEFAULT_firewall4=y
 CONFIG_DEFAULT_fstools=y
 CONFIG_MODULE_DEFAULT_fstools=y
-CONFIG_DEFAULT_iwinfo=y
-CONFIG_MODULE_DEFAULT_iwinfo=y
 CONFIG_MODULE_DEFAULT_kmod-ata-dwc=y
 CONFIG_DEFAULT_kmod-ath9k=y
 CONFIG_MODULE_DEFAULT_kmod-ath9k=y
@@ -134,8 +132,6 @@
 CONFIG_MODULE_DEFAULT_libc=y
 CONFIG_DEFAULT_libgcc=y
 CONFIG_MODULE_DEFAULT_libgcc=y
-CONFIG_DEFAULT_libustream-mbedtls=y
-CONFIG_MODULE_DEFAULT_libustream-mbedtls=y
 CONFIG_DEFAULT_logd=y
 CONFIG_MODULE_DEFAULT_logd=y
 CONFIG_DEFAULT_mtd=y
@@ -171,8 +167,6 @@
 CONFIG_MODULE_DEFAULT_urandom-seed=y
 CONFIG_DEFAULT_urngd=y
 CONFIG_MODULE_DEFAULT_urngd=y
-CONFIG_DEFAULT_wpad-basic-mbedtls=y
-CONFIG_MODULE_DEFAULT_wpad-basic-mbedtls=y
 CONFIG_HAS_TESTING_KERNEL=y
 CONFIG_HAS_FPU=y
 CONFIG_AUDIO_SUPPORT=y
@@ -186,6 +180,7 @@
 CONFIG_USES_INITRAMFS=y
 CONFIG_USES_SQUASHFS=y
 CONFIG_NAND_SUPPORT=y
+CONFIG_WIRELESS_SUPPORT=y
 CONFIG_powerpc=y
 CONFIG_ARCH="powerpc"
 
@@ -214,6 +209,7 @@
 # Root filesystem images
 #
 # CONFIG_TARGET_ROOTFS_EXT4FS is not set
+# CONFIG_TARGET_ROOTFS_JFFS2_NAND is not set
 CONFIG_TARGET_ROOTFS_SQUASHFS=y
 CONFIG_TARGET_SQUASHFS_BLOCK_SIZE=256
 CONFIG_TARGET_UBIFS_FREE_SPACE_FIXUP=y
@@ -251,6 +247,9 @@
 # CONFIG_IPK_FILES_CHECKSUMS is not set
 # CONFIG_REPRODUCIBLE_DEBUG_INFO is not set
 # CONFIG_COLLECT_KERNEL_DEBUG is not set
+CONFIG_TLS_PROVIDER_MBEDTLS=y
+# CONFIG_TLS_PROVIDER_OPENSSL is not set
+# CONFIG_TLS_PROVIDER_WOLFSSL is not set
 
 #
 # Kernel build options
```

  </details>


  <details>
    <summary><b>apm821xx/sata</b> config diff (<i>click to expand</i>)</summary>


```shell
diff -u config-master-apm821xx_sata config-tls-apm821xx_sata
```

```diff
--- config-master-apm821xx_sata	2023-07-28 16:52:20.119328886 +0200
+++ config-tls-apm821xx_sata	2023-07-28 16:52:26.285389934 +0200
@@ -91,7 +91,6 @@
 CONFIG_MODULE_DEFAULT_fdisk=y
 CONFIG_DEFAULT_fstools=y
 CONFIG_MODULE_DEFAULT_fstools=y
-CONFIG_MODULE_DEFAULT_iwinfo=y
 CONFIG_MODULE_DEFAULT_kmod-ata-dwc=y
 CONFIG_DEFAULT_kmod-dm=y
 CONFIG_MODULE_DEFAULT_kmod-dm=y
@@ -115,8 +114,6 @@
 CONFIG_MODULE_DEFAULT_libc=y
 CONFIG_DEFAULT_libgcc=y
 CONFIG_MODULE_DEFAULT_libgcc=y
-CONFIG_DEFAULT_libustream-mbedtls=y
-CONFIG_MODULE_DEFAULT_libustream-mbedtls=y
 CONFIG_DEFAULT_logd=y
 CONFIG_MODULE_DEFAULT_logd=y
 CONFIG_DEFAULT_lsblk=y
@@ -147,7 +144,6 @@
 CONFIG_MODULE_DEFAULT_urandom-seed=y
 CONFIG_DEFAULT_urngd=y
 CONFIG_MODULE_DEFAULT_urngd=y
-CONFIG_MODULE_DEFAULT_wpad-basic-mbedtls=y
 CONFIG_HAS_TESTING_KERNEL=y
 CONFIG_HAS_FPU=y
 CONFIG_AUDIO_SUPPORT=y
@@ -196,6 +192,7 @@
 # CONFIG_TARGET_EXT4_BLOCKSIZE_1K is not set
 CONFIG_TARGET_EXT4_BLOCKSIZE=4096
 # CONFIG_TARGET_EXT4_JOURNAL is not set
+# CONFIG_TARGET_ROOTFS_JFFS2_NAND is not set
 CONFIG_TARGET_ROOTFS_SQUASHFS=y
 CONFIG_TARGET_SQUASHFS_BLOCK_SIZE=256
 CONFIG_TARGET_UBIFS_FREE_SPACE_FIXUP=y
@@ -236,6 +233,9 @@
 # CONFIG_IPK_FILES_CHECKSUMS is not set
 # CONFIG_REPRODUCIBLE_DEBUG_INFO is not set
 # CONFIG_COLLECT_KERNEL_DEBUG is not set
+CONFIG_TLS_PROVIDER_MBEDTLS=y
+# CONFIG_TLS_PROVIDER_OPENSSL is not set
+# CONFIG_TLS_PROVIDER_WOLFSSL is not set
 
 #
 # Kernel build options
@@ -1689,7 +1689,6 @@
 #
 # Extra packages
 #
-CONFIG_PACKAGE_libiwinfo-data=m
 # end of Extra packages
 
 #
@@ -2892,7 +2891,7 @@
 # CONFIG_PACKAGE_libiconv-full is not set
 # CONFIG_PACKAGE_libintl-full is not set
 # CONFIG_PACKAGE_libiw is not set
-CONFIG_PACKAGE_libiwinfo=m
+# CONFIG_PACKAGE_libiwinfo is not set
 CONFIG_PACKAGE_libjson-c=y
 # CONFIG_PACKAGE_libltdl is not set
 # CONFIG_PACKAGE_liblua is not set
@@ -3136,16 +3135,12 @@
 # CONFIG_PACKAGE_hostapd-basic-mbedtls is not set
 # CONFIG_PACKAGE_hostapd-basic-openssl is not set
 # CONFIG_PACKAGE_hostapd-basic-wolfssl is not set
-CONFIG_PACKAGE_hostapd-common=m
+# CONFIG_PACKAGE_hostapd-common is not set
 # CONFIG_PACKAGE_hostapd-mbedtls is not set
 # CONFIG_PACKAGE_hostapd-mini is not set
 # CONFIG_PACKAGE_hostapd-openssl is not set
-# CONFIG_PACKAGE_hostapd-utils is not set
 # CONFIG_PACKAGE_hostapd-wolfssl is not set
-# CONFIG_PACKAGE_wpa-cli is not set
 # CONFIG_PACKAGE_wpa-supplicant is not set
-# CONFIG_WPA_RFKILL_SUPPORT is not set
-CONFIG_WPA_MSG_MIN_PRIORITY=3
 # CONFIG_WPA_WOLFSSL is not set
 # CONFIG_DRIVER_11AC_SUPPORT is not set
 # CONFIG_DRIVER_11AX_SUPPORT is not set
@@ -3158,7 +3153,7 @@
 # CONFIG_PACKAGE_wpa-supplicant-wolfssl is not set
 # CONFIG_PACKAGE_wpad is not set
 # CONFIG_PACKAGE_wpad-basic is not set
-CONFIG_PACKAGE_wpad-basic-mbedtls=m
+# CONFIG_PACKAGE_wpad-basic-mbedtls is not set
 # CONFIG_PACKAGE_wpad-basic-openssl is not set
 # CONFIG_PACKAGE_wpad-basic-wolfssl is not set
 # CONFIG_PACKAGE_wpad-mbedtls is not set
@@ -3342,7 +3337,7 @@
 # CONFIG_PACKAGE_iconv is not set
 # CONFIG_PACKAGE_ipcs is not set
 # CONFIG_PACKAGE_iwcap is not set
-CONFIG_PACKAGE_iwinfo=m
+# CONFIG_PACKAGE_iwinfo is not set
 CONFIG_PACKAGE_jshn=y
 CONFIG_PACKAGE_libjson-script=y
 # CONFIG_PACKAGE_libxml2-utils is not set
```

  </details>

kwiateusz commented:
Thx for gist, according to it some devices lost iwinfo from config ex. Ipq806, ipq807. Isn't that problem?

ynezz commented:
> Indeed, there are still some weird targets like apm821xx, lantiq, kirkwood, mediatek/mt7629 to name a few.

mediatek/mt7629 is fixed, its expected, that there is no wireless

Other mentioned targets shows a flaw in the current implementation, where I'm adding `DEFAULT_PACKAGES+=wpad-basic-mbedtls` and then trying to remove it via `PACKAGE_NO_WIRELESS:=-wpad-basic-mbedtls` in `DEVICE_PACKAGES:=-$(PACKAGE_NO_WIRELESS)` which of course doesn't work. Looking into this, I've also found, that we need to handle `uencrypt-mbedtls` in a same spirit (found in ath79 and mt7621) as well.

aparcar commented:
@ynezz I rebased the changes to have the CI running again. From my understanding this should still go into 23.03? If so @hauke please give this another try before creating rc4.

xabolcs commented:
> From my understanding this should still go into 23.03?

@aparcar, you ended up writing `23.03`, but you wanted to write `23.05` and not `22.03`, right?

hauke commented:
I do not want to merge this just some days before the final release, if this should go into 23.05 it should go in soon. Currently it looks like there is not much interest in this change.

ynezz commented:
It was marked as a draft for a reason, during testing I've found out, that it doesn't work, so it shouldn't be merged in the current state :)

> Currently it looks like there is not much interest in this change.

Its just lack of time, sorry. Anyway, if we want to fix that original issue #12874, we should probably consider some faster to implement approach. We could simply take the list of the targets [marked in the current PR as grand-flash targets](https://github.com/openwrt/openwrt/pull/12917/commits/c0721bca0330106333fe3daa7032f9f513e4247c)  (maybe extend it?) and convert them to `wpad-basic-openssl` somehow directly.

Djfe commented:
they just released the next lts release.
it contains full support for TLS 1.3
https://github.com/Mbed-TLS/mbedtls/releases/tag/v3.6.0

Djfe commented:
The next mbedTLS LTS release is being worked on
Follow #15092 for more
